File: solution/java/day12/src/main/java/greeting/Greeter.java
package greeting;

public interface Greeter {
    String CASUAL = "casual";
    String INTIMATE = "intimate";
    String FORMAL = "formal";

    static Greeter create() {
        return new Default();
    }

    static Greeter create(String formality) {
        return switch (formality) {
            case CASUAL -> new Casual();
            case INTIMATE -> new Intimate();
            case FORMAL -> new Formal();
            default -> create();
        };
    }

    default String greet() {
        return "Hello.";
    }

    class Default implements Greeter {
    }

    class Formal implements Greeter {
        @Override
        public String greet() {
            return "Good evening, sir.";
        }
    }

    class Casual implements Greeter {
        @Override
        public String greet() {
            return "Sup bro?";
        }
    }

    class Intimate implements Greeter {
        @Override
        public String greet() {
            return "Hello Darling!";
        }
    }
}

File: solution/c#/Day12/Day12/Greeter.cs
namespace Day12
{
    public static class GreeterFactory
    {
        public const string Casual = "casual";
        public const string Intimate = "intimate";
        public const string Formal = "formal";

        private static IGreeter Create() => new Default();

        public static IGreeter Create(string formality)
            => formality switch
            {
                Casual => new Casual(),
                Intimate => new Intimate(),
                Formal => new Formal(),
                _ => Create()
            };
    }

    public interface IGreeter
    {
        public string Greet() => "Hello.";
    }

    public class Default : IGreeter;

    public class Formal : IGreeter
    {
        public string Greet() => "Good evening, sir.";
    }

    public class Casual : IGreeter
    {
        public string Greet() => "Sup bro?";
    }

    public class Intimate : IGreeter
    {
        public string Greet() => "Hello Darling!";
    }
}